File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.js
function solution(maps) {
    // 우하좌상
    const dr = [0, 1, 0, -1];
    const dc = [1, 0, -1, 0];
    
    const mrows = maps.length;
    const mcols = maps[0].length;
    
    const queue = [[[0, 0], 1]];
    maps[0][0] = 0;
    
    while (queue.length>0) {
        const [[r, c], cnt] = queue.shift();
        
        if (r===mrows-1 && c===mcols-1) return cnt;
        
        for (let i=0; i<4; i++) {
            const nr = r + dr[i];
            const nc = c + dc[i];
            if (nr>=0 && nr<mrows && nc>=0 && nc<mcols && maps[nr][nc]) {
                queue.push([[nr, nc], cnt+1]);
                maps[nr][nc] = 0;
            }
        }
    }
    
    return -1;
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque
def solution(maps):
    n = len(maps)
    m = len(maps[0])
    
    q = deque()
    q.append(([0,0],1))
    maps[0][0] = 0
    # 우하좌상
    dir = [[0,1], [1,0], [0,-1], [-1,0]]
    
    while q:
        [r,c], cnt = q.popleft()
        if r==n-1 and c==m-1:
            return cnt
        for i in range(4):
            dr = r+dir[i][0]
            dc = c+dir[i][1]
            if dr<0 or dr>=n or dc<0 or dc>=m or maps[dr][dc]==0:
                continue
            else:
                q.append(([dr,dc], cnt+1))
                maps[dr][dc]=0
    return -1